File: GED.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar 23 16:18:14 2017

@author: hp
"""

import csv
import re
#Open Score Table
csv_reader = csv.reader(open('ScoreTable1.csv'))
ScoreTable = [[0 for col in range(26)] for row in range(26)]

for row in csv_reader:

    for list in ScoreTable:
        for s in range(0,26):
            list[s] = row[s]
        

def Max(a, b, c):
    max=a
    if max<b:
        max=b
    if max<c:
        max=c
    return max
    
def GetIndex(character):
    index = 0
    if character=='a':
        index = 0
    elif character=='b':
        index = 1
    elif character=='c':
        index = 2
    elif character=='d':
        index = 3
    elif character=='e':
        index = 4
    elif character=='f':
        index = 5
    elif character=='g':
        index = 6
    elif character=='h':
        index = 7
    elif character=='i':
        index = 8
    elif character=='j':
        index = 9
    elif character=='k':
        index = 10
    elif character=='l':
        index = 11
    elif character=='m':
        index = 12
    elif character=='n':
        index = 13
    elif character=='o':
        index = 14
    elif character=='p':
        index = 15
    elif character=='q':
        index = 16
    elif character=='r':
        index = 17
    elif character=='s':
        index = 18
    elif character=='t':
        index = 19
    elif character=='u':
        index = 20
    elif character=='v':
        index = 21
    elif character=='w':
        index = 22
    elif character=='x':
        index = 23
    elif character=='y':
        index = 24
    elif character=='z':
        index = 25
    return index

def GetReplaceScore(q, d):
    que_index = int(GetIndex(q))
    dic_index = int(GetIndex(d))
    return ScoreTable[que_index][dic_index]

def GED_Levenshtein(QueString, DicString):
    insert = -1
    delete = -1
    replace = -1
    match = 1
    scoreMatrix = [[0 for col in range(len(QueString)+1)] for row in range(len(DicString)+1)]
    # Initialize scoring matrix
    for DicLetter in range(len(DicString)+1):
        scoreMatrix[DicLetter][0]=-1*DicLetter
    for QueLetter in range(len(QueString)+1):
        scoreMatrix[0][QueLetter]=-1*QueLetter
    
    for d in range(1, len(DicString)+1):
        for q in range(1, len(QueString)+1):
            if DicString[d-1]!=QueString[q-1]:
                insertScore = scoreMatrix[d][q-1]+insert
                deleteScore = scoreMatrix[d-1][q]+delete
                replaceScore = scoreMatrix[d-1][q-1]+replace
                scoreMatrix[d][q] = Max(insertScore, deleteScore, replaceScore)
            elif DicString[d-1]==QueString[q-1]:
                scoreMatrix[d][q] = scoreMatrix[d-1][q-1]+match
                
    return scoreMatrix[len(DicString)][len(QueString)]

            
def GED_Soundex(QueString, DicString):
    insert = -3
    delete = -3
    match = 3
    scoreMatrix = [[0 for col in range(len(QueString)+1)] for row in range(len(DicString)+1)]
    # Initialize scoring matrix
    for DicLetter in range(len(DicString)+1):
        scoreMatrix[DicLetter][0]=-1*DicLetter
    for QueLetter in range(len(QueString)+1):
        scoreMatrix[0][QueLetter]=-1*QueLetter
    
    for d in range(1, len(DicString)+1):
        for q in range(1, len(QueString)+1):
            if DicString[d-1]!=QueString[q-1]:
                insertScore = scoreMatrix[d][q-1]+insert
                deleteScore = scoreMatrix[d-1][q]+delete
                replaceScore = scoreMatrix[d-1][q-1]+int(GetReplaceScore(QueString[q-1],DicString[d-1]))
                scoreMatrix[d][q] = Max(insertScore, deleteScore, replaceScore)
            elif DicString[d-1]==QueString[q-1]:
                scoreMatrix[d][q] = scoreMatrix[d-1][q-1]+match
                
    return scoreMatrix[len(DicString)][len(QueString)]  
    
#try1 = open('try1.txt', 'r')
dictionary = open('names.txt', 'r')
expect = open('expect.txt','w')
#l_expect = open('l_expect.txt', 'w')
#word = try1.readlines()
entries = dictionary.readlines()

train = open('train.txt','r')
correctAnswers = train.readlines()

row_index = 0
answerlist = [0]*len(correctAnswers)
testlist = [0]*len(correctAnswers)

#Get queryword and the answer from train.txt
for rows in correctAnswers:
    eachrow = re.split('\t', rows)
    answerlist[row_index] = eachrow[1]
    testlist[row_index] = eachrow[0]
    row_index = row_index+1
    
"""
#use lev to calculate the rate
count = 0
for queryWord in testlist:
    best = -100
    queryWord = queryWord.lower()
    for entry in entries:
        score = GED_Levenshtein(queryWord, entry)
        if score>best:
            best = score
            result = entry
    l_expect.write(result)
    print ("write "+str(count)+" times.")
    count=count+1;
    
l_expect.close()

num = 0
n=0
expectAnswers = open('l_expect.txt','r')
for expectAnswer in expectAnswers:
    n=n+1  
    if expectAnswer==answerlist[n-1]:
        num = num+1
      
l_expect.close()
"""  
#use soundex to calculate the rate
count = 0
for queryWord in testlist:
    best = -100
    queryWord = queryWord.lower()
    for entry in entries:
        score = GED_Soundex(queryWord, entry)
        if score>best:
            best = score
            result = entry
    expect.write(result)
    print ("write "+str(count)+" times.")
    count=count+1;
    
expect.close()

num = 0
n=0
expectAnswers = open('expect.txt','r')
for expectAnswer in expectAnswers:
    n=n+1
    if expectAnswer==answerlist[n-1]:
        num = num+1
expect.close()

#Close files
train.close()
dictionary.close()


    
rate = num/n
print (num)
print (n)
print ('%.2f%%' % (rate*100))
expectAnswers.close
    

File: Global Distance/src/GED.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;


public class GED {
	static FileReader train;
	static FileReader dictionary;
	static List<String> queryWords = new ArrayList<String>();
	static List<String> answers = new ArrayList<String>();
	static List<String> entries = new ArrayList<String>(); 
	public static void main(String[] args) {		
		try {
			train = new FileReader("C:/Users/hp/Desktop/Knowledge Technology/2017S1-90049P1-data-dos/train.txt");
			dictionary = new FileReader("C:/Users/hp/Desktop/Knowledge Technology/2017S1-90049P1-data-dos/names.txt");
			BufferedReader reader = new BufferedReader(train);
			BufferedReader dicReader = new BufferedReader(dictionary);
			
			File resultList = new File("result4.txt");
			FileWriter filewriter = new FileWriter(resultList.getName());
			BufferedWriter bufferedwriter = new BufferedWriter(filewriter);
			FileReader resultfile = new FileReader("result4.txt");
			BufferedReader read_result = new BufferedReader(resultfile);
			
			String row = null;
			String[] getSplit = null;
			String queryWord = null;
			String answer = null;
			String entry = null;
			while((row = reader.readLine())!=null){
				getSplit = row.split("\t");
				queryWord = getSplit[0];
				queryWord = queryWord.toLowerCase();
				answer = getSplit[1];
				queryWords.add(queryWord);
				answers.add(answer);
			}
			while((row = dicReader.readLine())!=null){
				entries.add(row);
			}
			int count = 0;
			for(String word:queryWords){
				int bestScore = -100;
				String result = null;
				for(String entryWord:entries){
					int temp_score = GED_Soundex(word, entryWord);
					if(temp_score>bestScore){
						bestScore = temp_score;
						result = entryWord;
					}
				}
				bufferedwriter.write(result+"\n");	
				System.out.println("Write "+count+" times.");
				count++;
			}
			bufferedwriter.close();
			/*calculate the correct rate*/
			String expect_result;
			int correctNum = 0;
			int total = 13437;
			double correctRate = 0;
			List result_list = new ArrayList<>();
			while((expect_result=read_result.readLine())!=null){
				result_list.add(expect_result);
			}
			for(int i=0;i<result_list.size();i++){
				if(result_list.get(i).equals(answers.get(i))){
					correctNum = correctNum+1;
				}
			}
			correctRate = (double)correctNum/total;
			DecimalFormat df = new DecimalFormat("0.00%");
			System.out.println("correct: "+correctNum);
			System.out.println("total: "+total);
			System.out.println(df.format(correctRate));

		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

	}
	public static int Max(int a, int b, int c){
		int max = a;
		if (max<b)
	        max=b;
	    if (max<c)
	        max=c;
	    return max;
	}

	public static int getReplaceScore(String q, String e){
		int score = 0;
		if(q.equals(e)){
			score = 2;//1
		}
		else{
			if(q.equals("a")){
				switch(e){
				case "y":
					score = 0;
					break;
				case "e":
					score = 0;//1
					break;
				case "i":
					score = -1;
					break;
				case "o":
					score = -1;
					break;
				case "u":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("e")){
				switch(e){
				case "a":
					score = 0;//1
					break;
				case "i":
					score = -1;
					break;
				case "o":
					score = -1;
					break;
				case "u":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("i")){
				switch(e){
				case "a":
					score = -1;
					break;
				case "e":
					score = -1;
					break;
				case "o":
					score = -1;
					break;
				case "u":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("o")){
				switch(e){
				case "a":
					score = -1;
					break;
				case "e":
					score = -1;
					break;
				case "i":
					score = -1;
					break;
				case "u":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("u")){
				switch(e){
				case "a":
					score = -1;
					break;
				case "e":
					score = -1;
					break;
				case "i":
					score = -1;
					break;
				case "o":
					score = -1;
					break;
				case "v":
					score = 1;//0
					break;
				case "w":
					score = 1;//0
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("y")){
				switch(e){
				case "a":
					score = 0;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("w")){
				switch(e){
				case "v":
					score = 1;//0
					break;
				case "u":
					score = 1;//0
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("v")){
				switch(e){
				case "w":
					score = 1;//0
					break;
				case "u":
					score = 1;//0
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("m")){
				switch(e){
				case "n":
					score = -1;
					break;
				default:
					score = -2; 
					break;
				}
			}
			else if(q.equals("n")){
				switch(e){
				case "m":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("d")){
				switch(e){
				case "t":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("t")){
				switch(e){
				case "d":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("b")){
				switch(e){
				case "p":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else if(q.equals("p")){
				switch(e){
				case "b":
					score = -1;
					break;
				default:
					score = -2;
					break;
				}
			}
			else{
				score = -2;
			}
		}
		return score;
	}
	
	public static int GED_Soundex(String QueString, String DicString){
		int insert = -1;
		int delete = -1;
		int[][] scoreMatrix = new int[DicString.length()+1][QueString.length()+1];
		/*Initialize scoring matrix*/
		for(int i=0;i<=DicString.length();i++){
			scoreMatrix[i][0] = i*-1;
		}
		for(int j=0;j<=QueString.length();j++){
			scoreMatrix[0][j] = j*-1;
		}
		/*Calculate the score*/
		String[] Que = QueString.split("");
		String[] Dic = DicString.split("");
		for(int q=1;q<=Que.length;q++){
			for(int d=1;d<=Dic.length;d++){
				if(Que[q-1].equals(Dic[d-1])==false){
					int insertScore = scoreMatrix[d-1][q]+insert;
					int deleteScore = scoreMatrix[d][q-1]+delete;
					int replaceScore = scoreMatrix[d-1][q-1]+getReplaceScore(Que[q-1],Dic[d-1]);
					scoreMatrix[d][q] = Max(insertScore, deleteScore, replaceScore);
				}
				else if(Que[q-1].equals(Dic[d-1])){
					scoreMatrix[d][q] = scoreMatrix[d-1][q-1]+getReplaceScore(Que[q-1],Dic[d-1]);
				}
			}
		}
		/*for(int i=0;i<scoreMatrix.length;i++){
			for(int j=0;j<scoreMatrix[i].length;j++){
				System.out.print(scoreMatrix[i][j]);
			}
			System.out.println();
		}*/	
		return scoreMatrix[DicString.length()][QueString.length()];
	}
}
